electron/new_config.ts

```diff
@@ -0,0 +1,294 @@
+        return 'done'
+      })
+    }
+
+    if (!existsSync(`${heroicToolsPath}/proton`)) {
+      exec(`mkdir '${heroicToolsPath}/proton' -p`, () => {
+        return 'done'
+      })
+    }
+
+    const protonPaths: string[] = [`${heroicToolsPath}/proton/`]
+    const foundPaths = steamPaths.filter((path) => existsSync(path))
+
+    const defaultWine = { bin: '', name: '' }
+    await execAsync(`which wine`)
+      .then(async ({ stdout }) => {
+        defaultWine.bin = stdout.split('\n')[0]
+        const { stdout: out } = await execAsync(`wine --version`)
+        defaultWine.name = `Wine - ${out.split('\n')[0]}`
+      })
+      .catch(() => console.log('Wine not installed'))
+
+    foundPaths.forEach((path) => {
+      protonPaths.push(`${path}/steamapps/common/`)
+      protonPaths.push(`${path}/compatibilitytools.d/`)
+      return
+    })
+
+    const lutrisPath = `${home}/.local/share/lutris`
+    const lutrisCompatPath = `${lutrisPath}/runners/wine/`
+    const proton: Set<{ bin: string; name: string }> = new Set()
+    const altWine: Set<{ bin: string; name: string }> = new Set()
+    const customPaths: Set<{ bin: string; name: string }> = new Set()
+
+    protonPaths.forEach((path) => {
+      if (existsSync(path)) {
+        readdirSync(path).forEach((version) => {
+          if (version.toLowerCase().startsWith('proton')) {
+            proton.add({
+              bin: `'${path}${version}/proton'`,
+              name: `Proton - ${version}`
+            })
+          }
+        })
+      }
+    })
+
+    if (existsSync(lutrisCompatPath)) {
+      readdirSync(lutrisCompatPath).forEach((version) => {
+        altWine.add({
+          bin: `'${lutrisCompatPath}${version}/bin/wine64'`,
+          name: `Wine - ${version}`
+        })
+      })
+    }
+
+    readdirSync(`${heroicToolsPath}/wine/`).forEach((version) => {
+      altWine.add({
+        bin: `'${lutrisCompatPath}${version}/bin/wine64'`,
+        name: `Wine - ${version}`
+      })
+    })
+
+    // skips this on new installations to avoid infinite loops
+    if (existsSync(heroicConfigPath)) {
+      const { customWinePaths } = await this.getSettings()
+      if (customWinePaths.length) {
+        customWinePaths.forEach((path: string) => {
+          if (path.endsWith('proton')) {
+            return customPaths.add({
+              bin: `'${path}'`,
+              name: `Proton Custom - ${path}`
+            })
+          }
+          return customPaths.add({
+            bin: `'${path}'`,
+            name: `Wine Custom - ${path}`
+          })
+        })
+      }
+    }
+
+    return [defaultWine, ...altWine, ...proton, ...customPaths]
+  }
+
+  public isLoggedIn() {
+    return existsSync(userInfo)
+  }
+
+  public getUserInfo() {
+    if (this.isLoggedIn()) {
+      return JSON.parse(readFileSync(userInfo, 'utf-8'))
+    }
+    return { account_id: '', displayName: null }
+  }
+
+  protected abstract getSettings() : Promise<AppSettings>
```

Comment:
That's the first time ever I saw anyone using `protected` or `abstract` in Javascript.
I'll need to do some research on how this features is implemented and how it works because, as far as I know, `private` and `public` are experimental, not approved by ECMA so far, but it is 'supported' on a experimental stage in chrome and should work on Electron. BUT protected and abstract I'm not so sure.

I'm just afraid that this can lead to unexpected behavior and more bugs that we can think that might be related to the code but it is actually a JS issue.

I know that you want to make this looks nice and all, using OOP, but unfortunately, JS is not an OOP language, it wasn't created for that and most of the things we have, like using TS to improve this compatibility, are workarounds.

Even classes are no more than objects in JS.

so let's use it with care. In my opinion, it is ok if you want to use classes, but I don't really know if the other features will work as you might think they will, like on C++ or Java, for instance.

Reply:
I tested and it looks to be working just fine. These are standard JS constructs, don't know why they're not used as often.

>Even classes are no more than objects in JS.
Not after ES6. And isn't Heroic ES6?

Reply:
Classes are ES6 but not private, public, abstract, protected, etc.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes

Reply:
Better: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields
first square :D

Reply:
As I see, it should work well with Electron v11 or higher. Also access modifiers can be dropped, I just use them since its broadly supported.